hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java

```diff
@@ -239,7 +293,15 @@ private void handleParityWrites(int parityCellSize,
     // TODO: we should alter the put block calls to share CRC to each stream.
     ECBlockOutputStreamEntry streamEntry =
         blockOutputStreamEntryPool.getCurrentStreamEntry();
+    // Since writes are async, let's check the failures once.
+    if(streamEntry.checkStreamFailures()){
```

Comment:
With this call, we are synchronising on parity write, and we loose the async nature of the write on this front as well. We discussed today that the grpc client does not do writes async, if we fix that, then we still loose async writes as parity writes become synced. I believe this we should record as a future improvement here, as handling the rewrite of the stripe in a more async manner would probably require a significant change in how we buffer data.

Reply:
The problem is we cannot allow the writer asynchronously out of one full stripe. Otherwise we have to cache all the stripe data which could be costly.  Let's say we allow multiple stripes asynchronously and later after writing multiple stripes data, first stripe received failure. If we don't cache all of the stripes we have written so far, we can;t copy back that data to new blockgroup. I think separately we can explore to cache elastically some stripes data and allow asynchronously. Currently the async nature applies to stripe level. I checked HDFS and we are inline with that behavior. Right after parity cells writing, we check for failures.

Reply:
Yes, that's what I meant as well, but I realized I forgot my point here :D I wanted to suggest to note this behaviour in the comment above the if or in checkStreamFailures API doc ;)